subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/TaskCacheKeyCalculator.java

```diff
@@ -0,0 +1,128 @@
+        BuildCacheKeyBuilder builder = new DefaultBuildCacheKeyBuilder();
+
+        builder.putString(execution.getTaskClass());
+        builder.putBytes(taskClassLoaderHash);
+        builder.putBytes(taskActionsClassLoaderHash);
+
+        // TODO:LPTR Use sorted maps instead of explicitly sorting entries here
+
+        for (Map.Entry<String, Object> entry : sortEntries(execution.getInputProperties().entrySet())) {
+            Object value = entry.getValue();
+            byte[] hash = hashForObject(value).asBytes();
+            addInputProperty(entry.getKey(), hash, builder, inputHashes);
+        }
+
+        for (Map.Entry<String, FileCollectionSnapshot> entry : sortEntries(execution.getInputFilesSnapshot().entrySet())) {
+            FileCollectionSnapshot snapshot = entry.getValue();
+            DefaultBuildCacheKeyBuilder newBuilder = new DefaultBuildCacheKeyBuilder();
+            snapshot.appendToCacheKey(newBuilder);
+            byte[] hash = newBuilder.buildHashCode().asBytes();
+            addInputProperty(entry.getKey(), hash, builder, inputHashes);
+        }
+
+        Set<String> outputPropertyNames = new HashSet<String>();
+        for (String cacheableOutputPropertyName : sortStrings(execution.getOutputPropertyNamesForCacheKey())) {
+            outputPropertyNames.add(cacheableOutputPropertyName);
+            builder.putString(cacheableOutputPropertyName);
+        }
+
+        BuildCacheKey cacheKey = builder.build();
+        if (hashesListener != null) {
+            hashesListener.inputsCollected(task, cacheKey, new TaskCachingHashesListener.TaskCachingInputs(
+                taskClassLoaderHash,
+                taskActionsClassLoaderHash,
+                inputHashes,
+                outputPropertyNames
+            ));
+        }
+        return cacheKey;
+    }
+
+    private static void addInputProperty(String name, byte[] hash, BuildCacheKeyBuilder builder, Map<String, byte[]> inputHashes) {
+        builder.putString(name);
+        builder.putBytes(hash);
+        inputHashes.put(name, hash);
+    }
+
+    private static <T> List<Map.Entry<String, T>> sortEntries(Set<Map.Entry<String, T>> entries) {
+        List<Map.Entry<String, T>> sortedEntries = Lists.newArrayList(entries);
+        Collections.sort(sortedEntries, new Comparator<Map.Entry<String, T>>() {
+            @Override
+            public int compare(Map.Entry<String, T> o1, Map.Entry<String, T> o2) {
+                return o1.getKey().compareTo(o2.getKey());
+            }
+        });
+        return sortedEntries;
+    }
+
+    private static List<String> sortStrings(Collection<String> entries) {
+        List<String> sortedEntries = Lists.newArrayList(entries);
+        Collections.sort(sortedEntries);
+        return sortedEntries;
+    }
+
+    private static HashCode hashForObject(Object object) {
+        return ((DefaultBuildCacheKeyBuilder) new DefaultBuildCacheKeyBuilder().appendToCacheKey(object)).buildHashCode();
```

Comment:
Could we change `appendToCacheKey` in `DefaultBuildCacheKeyBuilder` to just be a static method that we can call (that returns a hash)?